src/py/flwr/client/node_state.py

```diff
@@ -15,30 +15,55 @@
 """Node state."""
 
 
+from dataclasses import dataclass
+from pathlib import Path
 from typing import Any, Dict, Optional
 
 from flwr.common import Context, RecordSet
+from flwr.common.config import get_fused_config
+from flwr.common.typing import Run
+
+
+@dataclass()
+class RunInfo:
+    """Contains the Context and initial run_config of a Run."""
+
+    context: Context
+    initial_run_config: Dict[str, str]
 
 
 class NodeState:
     """State of a node where client nodes execute runs."""
 
     def __init__(self, partition_id: Optional[int]) -> None:
         self._meta: Dict[str, Any] = {}  # holds metadata about the node
-        self.run_contexts: Dict[int, Context] = {}
+        self.run_info_dict: Dict[int, RunInfo] = {}
         self._partition_id = partition_id
 
-    def register_context(self, run_id: int) -> None:
+    def register_context(
+        self,
+        run_id: int,
+        run_info: Optional[Run] = None,
```

Comment:
Should we rename this to avoid confusion with the new `RunInfo`?
```suggestion
        run: Optional[Run] = None,
```